tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -680,21 +709,67 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
       }
       return transp_dims;
     };
+    
+    // Plain transpose on a or b. Plain transposes a matrix by permuting its 
+    // dimension without changing storage order.
+    auto plain_transpose = [&](const char matrix_name) {
+      if (matrix_name == 'a') {
+        std::vector<int64_t> new_dim_order =
+            transp_dim_order(a, a_contracting_dims[0], a_batch_dims);
+        a = instr->AddInstruction(HloInstruction::CreateTranspose(
+            ShapeUtil::MakeShapeWithDenseLayout(
+                a->shape().element_type(), transp_dims(a, new_dim_order),
+                a->shape().layout().minor_to_major()),
+            a, new_dim_order));
+      } else if (matrix_name == 'b') {
+        std::vector<int64_t> new_dim_order =
+            transp_dim_order(b, b_contracting_dims[0], b_batch_dims);
+        b = instr->AddInstruction(HloInstruction::CreateTranspose(
+            ShapeUtil::MakeShapeWithDenseLayout(
+                b->shape().element_type(), transp_dims(b, new_dim_order),
+                b->shape().layout().minor_to_major()),
+            b, new_dim_order));
+      } else {
+        return InternalError("Invalid matrix name.");
+      }
+      return OkStatus();
+    };
 
-    if (is_col_major) {
-      std::vector<int64_t> new_dim_order =
-          transp_dim_order(a, a_contracting_dims[0], a_batch_dims);
-      a = instr->AddInstruction(HloInstruction::CreateTranspose(
-          ShapeUtil::MakeShape(a->shape().element_type(),
-                               transp_dims(a, new_dim_order)),
-          a, new_dim_order));
+    DotDimensionNumbers *dim_nums =
+        gemm_backend_config.mutable_dot_dimension_numbers();
+    // Apply necessary transposes to accommodate canonicalize matmul(lhs and rhs
+    // contracting dims are 1 and 0). Also assuming transpose folding pass later
+    // will remove duplcated transposes. The last transpose is required by 
+    // cublas fp8 matmul restriction.
```

Comment:
Please confirm my understanding of what the issue. IIUC, right now (without this PR), gemm_rewriter transposes the LHS of the matmul. This is to negate the tranpose that [`MatmulPlan` does](https://github.com/tensorflow/tensorflow/blob/4aec415b3f06b19c380d1a0ca92cc2de0d74cc21/tensorflow/compiler/xla/service/gpu/matmul_utils.cc#L812) to meet the cuBLAS requirement that the LHS [must be transposed](https://docs.nvidia.com/cuda/cublas/#cublasltmatmul).

The issue with this approach (that  this PR is addressing), is that this process doesn't take into account the layouts of the LHS and the RHS (although it does check the layout of the output). According to [this Github comment](https://github.com/tensorflow/tensorflow/pull/58720#discussion_r1044950772)

> The `cublasLtOrder_t` is provided for convenience. Actual matrix layout is figured out internally by also taking `cublasOperation_t` into account. So you can construct multiple valid combinations that all will work as long as they map to column-major TN.

The code currently assumes that the LHS must be transposed and the RHS must not be transposed, but in reality, you must also take into account the order (row-major vs column-major).

I'm still trying to understand this PR, but perhaps an alternative way to fix this would be the following: Remove the transpose in `Matmul` plan, then ensure the LHS is "effectively" row-major and the RHS is "effectively" column-major ("effectively" in this case meaning what `MatrixIsColumnMajor` returns as opposed to what the layout indicates, which can be different depending on the contracting dimensions). You can do by using `MatrixIsColumnMajor` to determine the order of the LHS and RHS, and tranposing the LHS if the LHS is not row-major and transposing the RHS if the RHS is not column major. This would avoid the large number of if-statements below.

Also can you add a test to `tensorflow/compiler/xla/tests/dot_operation_test.cc`, similar to the existing [FP8 tests](https://github.com/tensorflow/tensorflow/blob/2bce366956049b6db98f91b862529321a4a5046f/tensorflow/compiler/xla/tests/dot_operation_test.cc#L755)?

Reply:
See https://github.com/tensorflow/tensorflow/pull/59515#issuecomment-1419934401